src/Hamiltonians/Molecular.jl

```diff
@@ -0,0 +1,613 @@
+"""
+    MolecularMolecularHamiltonian(fcidump::String[, starting_address::FermiFS2C])
+    
+    MolecularMolecularHamiltonian(fd::ElemCo.FciDumps.FDump[, starting_address::FermiFS2C])
+
+Implements a Molecular Hamiltonian with [ElemCo.jl](https://elem.co.il). 
+
+The expression of Hamiltonian under Bohn-Oppenheimer approximation is 
+```math
+\\begin{aligned}
+    \\hat{H}   & = \\hat{H}_0 + \\hat{H}_1 + \\hat{H}_2, \\\\
+    \\hat{H}_1 & = \\sum_{i,j} h_{ij} a^{\\dagger}_j a_i, \\\\
+    \\hat{H}_2 & = \\sum_{kl,ij} V_{kl,ij} a^{\\dagger}_{k} a^\\dagger_{l} a_j a_i.
+\\end{aligned}
+```
+
+# Arguments
+* `fcidump`: The path to FCIDUMP file of molecular.
+* `fd`: The FCIDUMP structure defined in [ElemCo.jl](https://elem.co.il).
+* `starting_address`: The starting address, defines number of particles and sites.
+
+See also [`FermiFS2C`](@ref)
+
+!!! warning
+    The FCIDUMP file is assumed to be a Restricted Hartree-Fock(RHF) Hamiltonian.
+"""
+struct MolecularHamiltonian{T,A<:FermiFS2C,D<:FDump} <: AbstractHamiltonian{T}
+    starting_address::A
+    fcidump::D
+end
+
+function MolecularHamiltonian(fcidump::String, starting_address::Union{Nothing,FermiFS2C}=nothing)
+    fd = read_fcidump(fcidump, Val(4))
+    MolecularHamiltonian(fd, starting_address)
+end
+
+function MolecularHamiltonian(fd::QFDump, starting_address::Union{Nothing,FermiFS2C}=nothing)
+    if !fd_exists(fd)
+        throw(ArgumentError("invalid input FCIDUMP file"))
+    end
+    n_orb = headvar(fd, "NORB", Int)
+    n_elec = headvar(fd, "NELEC", Int)
+    ms2 = headvar(fd, "MS2", Int)
+
+    if isnothing(n_orb) || isnothing(n_elec) || isnothing(ms2)
+        throw(ArgumentError("input FCIDUMP file must have `NORB`, `NELEC`, and `MS2` defined in header"))
+    end
+
+    n_alpha_elec = (n_elec + ms2) ÷ 2
+    n_beta_elec = (n_elec - ms2) ÷ 2
+
+    if starting_address === nothing
+        starting_address = FermiFS2C(
+            near_uniform(FermiFS{n_alpha_elec,n_orb}),
+            near_uniform(FermiFS{n_beta_elec,n_orb})
+        )
+    end
+
+    if num_modes(starting_address) != n_orb
+        throw(ArgumentError("starting_address provided must have same orbital numbers with FCIDUMP"))
+    end
+
+    val_type = typeof(fd.int0)
+    addr_type = typeof(starting_address)
+    fd_type = typeof(fd)
+    MolecularHamiltonian{val_type,addr_type,fd_type}(starting_address, fd)
+end
+
+struct Modes{OA,OB,UA,UB,T<:FermiFSIndex}
+    occupied::Tuple{ModeMap{OA,T},ModeMap{OB,T}}
+    unoccupied::Tuple{ModeMap{UA,T},ModeMap{UB,T}}
+end
```

Comment:
I suggest a rename here, as well as switching the order of the type parameters. The typeindex `TI` is meaningful and thus should go first (in case one wants to check what type of indices it works with). The others are dummys and less meaningful.

```suggestion
struct FermiFS2CModeMaps{TI<:FermiFSIndex,OA,OB,UA,UB}
    occupied::Tuple{ModeMap{OA,TI},ModeMap{OB,TI}}
    unoccupied::Tuple{ModeMap{UA,TI},ModeMap{UB,TI}}
end
```
This type should live in `BitStingAddresses/multicomponent.jl`.